src/site/_utils/build-partial.js

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2019 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const fs = require("fs").promises;
+const path = require("path");
+const postEleventyBuild = require("./post-eleventy-build");
+
+const suffixLength = "index.html".length;
+
+const writePartial = async (to, raw) => {
+  await fs.mkdir(path.dirname(to), {recursive: true});
+  await fs.writeFile(to, JSON.stringify(raw));
+};
+
+module.exports = function buildPartial() {
+  const work = [];
+
+  postEleventyBuild(async () => {
+    await Promise.all(work);
+  });
+
+  return function(content, page, collections, renderData = {}) {
+    if (!page.outputPath.endsWith("/index.html")) {
+      return content; // unexpected output format
+    }
+
+    // nb. matches the Meta component, used for title
+    const pageData = {
+      ...collections.all.find((item) => item.fileSlug === page.fileSlug).data,
+      ...renderData,
+    };
+
+    const outputPath =
+      page.outputPath.substr(0, page.outputPath.length - suffixLength) +
+      "index.json";
+
+    const partial = {
```

Comment:
I like that idea and I like the API. But I think I'll try to avoid conflating these two goals right now.

Also, interestingly, we don't cache the JSON right now for initial load (which sort of makes sense). It's not a big deal, but perhaps users would be surprised if we don't index that page—maybe we aggressively refetch it somewhere when the SW is installed.

Reply:
Keeping those goals separate sounds fine.

When you say "initial load," do you mean the first navigation request that's made **before** the service worker is installed? If so, yeah, I wouldn't worry about that too much from a Content Indexing point of view.

(Or by "initial load," do you mean every navigation request even **after** the service worker is installed? I don't think that's what you mean, but it would be great to clarify in case I'm misunderstanding how the service worker logic works.)

Reply:
I definitely mean the first case. The initial HTML, without Service Worker, doesn't sneak in the JSON partial. I'm glad it's not worth worrying about, thanks.